internals/overlord/logstate/manager.go

```diff
@@ -129,3 +165,27 @@ func (m *LogManager) Stop() {
 	}
 	wg.Wait()
 }
+
+// ServiceData holds the data for each service that the log manager needs to
+// keep track of.
+type ServiceData struct {
+	Buffer *servicelog.RingBuffer
+	Env    map[string]string
+}
+
+// unchanged checks if d pertains to the same running service as other.
+func (d *ServiceData) unchanged(other *ServiceData) bool {
```

Comment:
I've inlined this logic now.

> if the process changes (so not the same running service, in a way), but the buffer is the same, we can keep the puller untouched.

This comment made me realise: instead of having a ServiceStarted method on the gatherer, we should have EnvChanged and BufferChanged, which are only called if the relevant service datum has changed.